packages/vue-app/template/components/nuxt-hmr.vue

```diff
@@ -0,0 +1,103 @@
+<template>
+<transition appear>
+  <div class="nuxt__hmr" v-if="building">
+      <div>Building ({{ progress}}%)</div>
+  </div>
+</transition>
+</template>
+
+<script>
+import { setTimeout } from 'timers';
+export default {
+  data() {
+    return {
+      building: false,
+      progress: 0
+    }
+  },
+  mounted() {
+    if (typeof WebSocket === undefined) {
+      return // Unsupported
+    }
+    this.wsConnect('<%= router.base %>_loading/ws')
+  },
+  methods: {
+    wsConnect(path) {
+      if (path) {
+        const protocol = location.protocol === 'https:' ? 'wss' : 'ws'
+        this.wsURL = `${protocol}://${location.hostname}:${location.port}${path}`
+      }
+
+      this.ws = new WebSocket(this.wsURL)
+      this.ws.onclose = this.onWSClose.bind(this)
+      this.ws.onerror = this.onWSError.bind(this)
+      this.ws.onmessage = this.onWSMessage.bind(this)
+    },
+
+    wsReconnect(e) {
+      setTimeout(() => {
+        this.wsConnect()
+      }, 1000)
+    },
+
+
+    onWSClose(e) {
+      // https://tools.ietf.org/html/rfc6455#section-11.7
+      if (e.code !== 1000 && e.code !== 1005) {
+        // Unkown error
+        this.wsReconnect()
+      }
+    },
+
+    onWSError(error) {
+      if (error.code === 'ECONNREFUSED') {
+        this.wsReconnect(error)
+      }
+    },
+
+    onWSMessage(msg) {
+      let data = msg.data
+
+      try {
+        if (data[0] === '{') {
+          data = JSON.parse(data)
+        }
+      } catch (e) {}
```

Comment:
If data is not json, is following code supposed to still continue executing?

The `allDone` and `states` are all `undefined` which will cause exception as well.